### PR TITLE
Requisitions paperwork reward

### LINKED ISSUE #2229: Requisitions crates should come with a stampable supply manifest that you can send back for some money

### PR DESCRIPTION
**Pull request by noisrevni-dev**
## About the PR
Add paperwork system for supply department to work on for extra money. As specified on the title of #2229 

- resolves #2229 

## Technical details
Its start with function for printing and attaching invoice to label slot of each crates. The invoice paper itself contains information about the crate and its content. With some flavor text to fill empty space, its just random numbers and meaningless information i copy from a military crate reference images. Next, the function for scanning crate content for stamped invoice and accumulate the reward. It works on the inside or `labelSlot` of the crate, both can be filled at the same time. These two function were the main part of the reward system. The invoice printer function called on `RequisitionsSystem.SpawnOrders()`, while the scanner is on `RequisitionsSystem.Sell()` .

It also had a function to send feedback to the client side to let players knows that they actually doing the paperwork correctly. The feedback is in the form of audio ping followed by in game chat flavor text (like cloning machine) telling the players that "the other side" has receive their paperwork and granted "budget surplus" to the "operation". The way it works is, after the paperwork reward has been calculated, it will check if there's accumulated rewards from paperwork. If it has, it will send feedback to **the last interacted ASRS computer** by any players. In ideal case, that would be the player who call the elevator down. **The invoice should be stamped with** `CMStampApproved` **stamp only**. Upstream stamp will give error texture.

As for now, the reward is 300. Combined with crate, its 500. Feel free to change it on `RequisitionsInvoiceComponent.Reward` to balance the system.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

Closed crates with invoice:
Screenshot from 2024-06-12 17-28-30

Opened crates with invoice:
Screenshot from 2024-06-12 17-29-56

Sending stamped invoice in label slot :
[normal-operation.webm](https://github.com/CM-14/CM-14/assets/152494932/b3e70e9b-7612-43ae-a674-a346316aa47e)

Sending non-stamped invoice in label slot :
[wrong-operation.webm](https://github.com/CM-14/CM-14/assets/152494932/72a7b6ea-6068-40c3-bea1-4ddb6c133d7f)

Sending stamped invoice inside the crate (bulk) :
[alternative-operation.webm](https://github.com/CM-14/CM-14/assets/152494932/5ba7ea7c-f6d8-40f3-aecb-6ffc16def520)

**Changelog**

:cl: Noisrevni
- add: Added requisitions paperwork reward system, now supply crew can send stamped invoice to gain extra money for the ongoing operation.

### CODE REVIEW

**Comment by Vermidia** [comment]
Since all the paper label sprites are the same, you should do what I did when I made the paper label PR and just have a single crate_label.rsi applied to the base crate with
```
    - state: paper
      sprite: (whatever your sprite name is here)
      map: ["enum.PaperLabelVisuals.Layer"]
```

**Comment by noisrevni-dev** [comment]
Thanks for the tips. I will try it after sending other PR today. My problem when trying to make label sprite is some of the crates had extra component that should not be obstructed. Secure crate had lights to indicate locking state and their light placement are not in the same place like upstream crate. So, some of the label are different than the other, even its just 1 pixel difference.

**Comment by Vermidia** [comment]
Offset is something you can also do with this system. You can see it upstream on the pirate chest and such.

**Comment by QuietlyWhisper** [comment]
It would be preferred if you could do offsets like the above comment mentions, so we only have one sprite used. If not just use same sprite on all crates and hopefully someone else or ourselves can take care of it, no worries.

Sorry for inconvenience.

**Review by QuietlyWhisper — CHANGES_REQUESTED** [review]
It would be preferred if you could do offsets like the above comment mentions, so we only have one sprite used. If not just use same sprite on all crates and hopefully someone else or ourselves can take care of it, no worries.

Sorry for inconvenience.

**Comment by noisrevni-dev** [comment]
No problem, I will try using layer and offset.

File: Resources/Textures/_CM14/Interface/Paper/paper_heading_requisition_invoice.svg.96dpi.png
Comment (QuietlyWhisper) [inline Resources/Textures/_CM14/Interface/Paper/paper_heading_requisition_invoice.svg.96dpi.png:1]: This

File: Resources/Textures/_CM14/Interface/Paper/paper_heading_requisition_invoice.svg
Comment (QuietlyWhisper) [inline Resources/Textures/_CM14/Interface/Paper/paper_heading_requisition_invoice.svg:1]: This

**Review by QuietlyWhisper — CHANGES_REQUESTED** [review]
Can you add a meta.json for these two files? We can't use images with no license unfortunately.
Format is like this:
\- files: ["Filename.file"]
  license: "[CC-BY...](https://creativecommons.org/share-your-work/cclicenses/)"
  copyright: "Taken from _"
  source: "Link to content if applicable"
  
  After that Smug will look at the code when she gets the time, thanks!

**Comment by noisrevni-dev** [comment]
the upstream doesn't have meta.json on their Paper interface. But i see they use original inkscape svg save file with DCMI on RDF standard. Should i follow that or use meta.json? I have no problem to provide the original inkscape svg save file since i made it on inkscape too.

**Comment by QuietlyWhisper** [comment]
> the upstream doesn't have meta.json on their Paper interface. But i see they use original inkscape svg save file with DCMI on RDF standard. Should i follow that or use meta.json? I have no problem to provide the original inkscape svg save file since i made it on inkscape too.

It's only for license, it doesn't need to be used for sprite states. It was probably not done on upstream because they didn't notice it in review I think. I think standard is to use meta.json for images even if it is not needed for sprite states
It tells anyone using our codebase how they're allowed to use these images, and credits yourself as the owner/creator of the images.

File: Resources/Textures/_CM14/Interface/Paper/paper_heading_requisition_invoice.svg.96dpi.png
Comment (noisrevni-dev) [reply Resources/Textures/_CM14/Interface/Paper/paper_heading_requisition_invoice.svg.96dpi.png:1]: Done

**Review by DrSmugleaf — COMMENTED** [review]
Thanks for the paper changes!
Testing this in-game it seems like the invoices spawn outside the crates for some reason, when ordering crates. Tested with two ammo crates:
image

**Comment by noisrevni-dev** [comment]
Okay, i will try to fix it.

File: Content.Server/_RMC14/Requisitions/RequisitionsSystem.cs [hunk 1]
```diff
@@ -340,15 +387,26 @@ private bool Sell(Entity<RequisitionsElevatorComponent> elevator)
...
             }
 
             QueueDel(entity);
         }
 
+        if (bureaucraticRewards > 0)
+        {
+            SendUIFeedback(Loc.GetString("requisition-paperwork-reward-message", ("amount", bureaucraticRewards)));
+            Console.WriteLine($"Bureaucratic Reward ({bureaucraticRewards} spessos)");
+        }
```
Comment (QuietlyWhisper) [inline Content.Server/_RMC14/Requisitions/RequisitionsSystem.cs:405]:
```suggestion
        if (bureaucraticRewards > 0)
            SendUIFeedback(Loc.GetString("requisition-paperwork-reward-message", ("amount", bureaucraticRewards)));
```

File: Content.Server/_RMC14/Requisitions/RequisitionsSystem.cs [hunk 2]
```diff
@@ -23,6 +24,9 @@ public sealed class RequisitionsSystem : SharedRequisitionsSystem
     [Dependency] private readonly IGameTiming _timing = default!;
     [Dependency] private readonly PhysicsSystem _physics = default!;
     [Dependency] private readonly SharedTransformSystem _transform = default!;
+    [Dependency] private readonly ChatSystem _chatSystem = default!;
+
+    public const string PaperRequisitionInvoice = "CMPaperRequisitionInvoice";
```
Comment (QuietlyWhisper) [inline Content.Server/_RMC14/Requisitions/RequisitionsSystem.cs]:
```suggestion
    public const string PaperRequisitionInvoice = "RMCPaperRequisitionInvoice";
```

File: Content.Shared/_RMC14/Requisitions/Components/RequisitionsInvoiceComponent.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
+[Access(typeof(SharedRequisitionsSystem))]
+public sealed partial class RequisitionsInvoiceComponent : Component
+{
+    [DataField, AutoNetworkedField]
+    public int Reward = 300;
+
+    [DataField]
+    public EntProtoId PaperOutput = "CMPaperRequisitionInvoice";
```
Comment (QuietlyWhisper) [inline Content.Shared/_RMC14/Requisitions/Components/RequisitionsInvoiceComponent.cs]:
```suggestion
    public EntProtoId PaperOutput = "RMCPaperRequisitionInvoice";
```

File: Resources/Prototypes/_RMC14/Entities/Objects/Misc/paper.yml [hunk 4]
```diff
@@ -41,3 +41,32 @@
   - type: Paper
     content: "Hey Garry, I got the boys to move most of the emergency supplies down into the ASRS hold just like ya' asked.
       Next time you're around Chinook I'll buy you a beer ok?"
+
+- type: entity
+  name: requisition invoice
+  parent: CMPaper
+  id: CMPaperRequisitionInvoice
```
Comment (QuietlyWhisper) [inline Resources/Prototypes/_RMC14/Entities/Objects/Misc/paper.yml]:
```suggestion
  id: RMCPaperRequisitionInvoice
```

**Review by QuietlyWhisper — APPROVED** [review]
Thanks

**Pull request merged**